python-sdk/nuscenes/eval/lidarseg/evaluate.py

```diff
@@ -0,0 +1,157 @@
+import os
+
+from tqdm import tqdm
+import numpy as np
+
+from nuscenes import NuScenes
+
+from utils import LidarsegChallengeAdaptor
+
+
+class LidarSegEval:
+    def __init__(self,
+                 nusc: NuScenes,
+                 results_folder: str,
+                 ignore_idx: int = 0):
+        """
+        """
+
+        """
+        # Check there are ground truth annotations.
+        self.gt_samples = nusc.sample # check if test set has samples?? Might have to check data itself.
+        assert len(self.samples) > 0, 'Error: There are no ground truth ______'
+
+        # Check results folder exists.
+        self.results_folder = results_folder
+        assert os.path.exists(results_folder), 'Error: The result folder ({}) does not exist!'.format(results_folder)
+        """
+
+        self.nusc = nusc
+        self.results_folder = results_folder
+
+        self.ignore_idx = ignore_idx
+
+        self.global_cm = None
+
+        self.adaptor = LidarsegChallengeAdaptor(self.nusc)
+
+        self.num_classes = len(self.adaptor.merged_name_2_merged_idx_mapping)
+        print('There are {} classes.'.format(self.num_classes))
+
+    def evaluate(self) -> None:
+        for sample in tqdm(self.nusc.sample):
+            sd_token = sample['data']['LIDAR_TOP']
+
+            # Load ground truth labels for sample data.
+            lidarseg_label_filename = os.path.join(self.nusc.dataroot,
+                                                   self.nusc.get('lidarseg', sd_token)['filename'])
+            lidarseg_label = self.load_bin_file(lidarseg_label_filename)
+
+            # print(lidarseg_label)
+            lidarseg_label = self.adaptor.convert_label(lidarseg_label)
+            # print(lidarseg_label)
+
+            # Load predictions for sample data.
+            lidarseg_pred_filename = os.path.join(self.results_folder, 'lidarseg',
+                                                  self.nusc.version, sd_token + '_lidarseg.bin')
+            lidarseg_pred = self.load_bin_file(lidarseg_pred_filename)
+
+            # TODO remove!!
+            lidarseg_pred = self.adaptor.convert_label(lidarseg_pred)
+
+            # Get the confusion matrix between the ground truth and predictions.
+            sd_cm = self._get_confusion_matrix(lidarseg_label, lidarseg_pred)
+
+            # Update the confusion matrix for the sample data into the confusion matrix for the split.
+            self._update_confusion_matrix(sd_cm)
+
+        """
+        # Can do some printing of each class IOU with the mapping
+        id2name = {label_id: label.name for label_id, label in res.labelmap.items() if label.name != 'ignore_label'}
+        """
+
+        iou_per_class = self._get_per_class_iou(self.global_cm)
```

Comment:
Add another metric for class agnostic IOU

Reply:
Added FWIOU